### PR TITLE
Leech pod monster family

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary

```SUMMARY: Content "Adds new leech plant monster family"```

#### Purpose of change
A new monster family that the player might encounter in still powered locations. Heavily inspired by the idea on this forum thread. 
https://discourse.cataclysmdda.org/t/new-monster-leech-pod-leech-stalk-leech-bloom-root-drones-area-power-plant/5982

#### Describe the solution

Basically there are 5 new foe types. All of them deal purely electrical damage, and all of them can drain power with a monster only EMP gun. Each monster has also a special role:

Leech stalks are area denial foes, they are immobile but can attack you with their emp gun from a range of 12 tiles maximum.

Leech pod clusters are foe spawners, constantly producing root drones and runners until there are 45 of them in the map. This growth limit ensures that the main leech pod plant area is saturated with foes, while keeping the rest of the reality bubble relatively free . The clusters eventually transform into leech stalks.

Leech blossom are a combination of both stalks and cluster, there's only one blossom active at any given time, but stalks can turn into new blossoms if the original is killed. Leech stalks far from the original blossom might also turn into new blossoms.

Both root drones and root runners are swarmer foes. They have a special enhanced attack, that if it manages to kill something, will turn them into either leech pod clusters or leech stalks, making them the only method by which the leech plant might expand. They can also attack you with their emp gun if they are adjacent.  Rather weak individually, but their electric damage and enhanced special attack makes them very dangerous in groups. 

Suggestion: Clear leech blooms using a flamethrower or explosives.

#### Describe alternatives you've considered

Principally, the leech pods could have their own emp effect, I however didnt consider this strictly necessary.

#### Testing

Theres a special microlab room that spawns these and some reactors. You can force spawn the room to check them out.

#### Additional context

Wip because I need to think of better descriptions (help with those is appreciated). And because I want some feedback on the balance of these things if possible

### CODE REVIEW

File: data/json/items/comestibles/carnivore.json [hunk 1]
```diff
@@ -932,5 +932,51 @@
...
     "volume": "250 ml",
     "fun": -30
+  },
+  {
+    "id": "meat_frond",
+    "copy-from": "meat_tainted",
+    "type": "COMESTIBLE",
+    "name": "alien fronds",
+    "use_action": "POISON",
+    "description": "The fleshy fronds harvested from an alien plant. The tiny membranous leaves attached and rigid and gut-like stem are certainly poisonous, and yet they have a paradoxically pleasant and inviting sweet smell.  Might be non-vegan.",
```
Comment (Night-Pryanik) [inline data/json/items/comestibles/carnivore.json]:
```suggestion
    "description": "The fleshy fronds harvested from an alien plant.  The tiny membranous leaves attached and rigid and gut-like stem are certainly poisonous, and yet they have a paradoxically pleasant and inviting sweet smell.  Might be non-vegan.",
```

While I understand the reason why the "...are certainly poisonous" part is in the description, but it seems kinda meta-gaming to me. How could player has known about this, especially if he didn't taste it before?

File: data/json/items/comestibles/carnivore.json [hunk 2]
```diff
@@ -932,5 +932,51 @@
...
+      ]
+    }
+  },
+  {
+    "id": "meat_bark",
+    "copy-from": "meat_frond",
+    "type": "COMESTIBLE",
+    "name": "leech bark",
+    "name_plural": "scraps of leech bark",
+    "description": "Dry and though bark matter harvested from an alien plant.  It is slightly translucent, and if placed against the light you can distinguish glistening blue veins running through it.",
```
Comment (Night-Pryanik) [inline data/json/items/comestibles/carnivore.json]:
```suggestion
    "description": "Dry and tough bark matter harvested from an alien plant.  It is slightly translucent, and if placed against the light you can distinguish glistening blue veins running through it.",
```

File: data/json/monsters/power_leech.json [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    "armor_bash": 15,
+    "vision_day": 30,
+    "vision_night": 30,
+    "luminance": 200,
+    "special_attacks": [
+      {
+        "type": "gun",
+        "cooldown": 5,
+        "gun_type": "emp_frond",
+        "fake_skills": [ [ "gun", 3 ], [ "rifle", 3 ] ],
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]: Rifle skill is useless here, as emp frond uses pistol skill? The same for other users of this gun.

File: data/json/monsters/power_leech.json [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    ],
+    "death_drops": {  },
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "NOHEAD", "IMMOBILE", "NO_BREATHE", "QUEEN", "HARDTOSHOOT" ]
+  },
+  {
+    "id": "mon_leech_stalk",
+    "type": "MONSTER",
+    "name": "leech stalk",
+    "description": "A resplendent and voluminous alien fern. A faint buzzing sound emanates from it, and the shadow cast by its canopy continuously glows with electric charge.",
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]:
```suggestion
    "description": "A resplendent and voluminous alien fern.  A faint buzzing sound emanates from it, and the shadow cast by its canopy continuously glows with electric charge.",
```

File: data/json/monsters/power_leech.json [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    ],
+    "death_drops": {  },
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "NOHEAD", "ELECTRONIC", "IMMOBILE", "NO_BREATHE", "HARDTOSHOOT" ]
+  },
+  {
+    "id": "mon_leech_pod_cluster",
+    "type": "MONSTER",
+    "name": "leech pod cluster",
+    "description": "The translucent egg pods of an alien plant, firmly attached by luminous rhizomes.  You can barely distinguish a root drone floating withing a cloudy substance.",
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]:
```suggestion
    "description": "The translucent egg pods of an alien plant, firmly attached by luminous rhizomes.  You can barely distinguish a root drone floating within a cloudy substance.",
```

File: data/json/monsters/power_leech.json [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    "hp": 40,
+    "speed": 100,
+    "material": [ "flesh" ],
+    "symbol": "g",
+    "color": "white",
+    "aggression": 100,
+    "morale": 100,
+    "armor_bash": 15,
+    "luminance": 60,
+    "starting_ammo": { "9mm": 100 },
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]:
```suggestion
```

File: data/json/monsters/power_leech.json [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    ],
+    "death_drops": {  },
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "NOHEAD", "IMMOBILE", "NO_BREATHE" ]
+  },
+  {
+    "id": "mon_leech_root_runner",
+    "type": "MONSTER",
+    "name": "root runner",
+    "description": "This clump of woody vegetation hastily clambers around in a lizard-like fashion.  Three translucent scale-leaves stand tall on the backside of the creature, and the thin ridges within them periodically glow through some unknown mean.  Its seemingly a symbiote of the nearby alien ferns, and looks ready to defend them with its life.",
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]:
```suggestion
    "description": "This clump of woody vegetation hastily clambers around in a lizard-like fashion.  Three translucent scale-leaves stand tall on the backside of the creature, and the thin ridges within them periodically glow through some unknown mean.  It's seemingly a symbiote of the nearby alien ferns, and looks ready to defend them with its life.",
```

File: data/json/monsters/power_leech.json [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    ],
+    "death_drops": {  },
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "NOHEAD", "ELECTRONIC", "NO_BREATHE", "HARDTOSHOOT" ]
+  },
+  {
+    "id": "mon_leech_root_drone",
+    "type": "MONSTER",
+    "name": "root drone",
+    "description": "A small bulb with a beak like protuberance, skittishly roaming about under three tendril rhizomes.  Dripping and glistening, it resembles a creature newly born rather than a sapling grown from seeds.",
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]:
```suggestion
    "description": "A small bulb with a beak-like protuberance, skittishly roaming about under three tendril rhizomes.  Dripping and glistening, it resembles a creature newly born rather than a sapling grown from seeds.",
```

File: data/json/monsters/power_leech.json [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      {
+        "type": "gun",
+        "cooldown": 15,
+        "gun_type": "emp_frond",
+        "fake_skills": [ [ "gun", 2 ], [ "rifle", 2 ] ],
+        "ranges": [ [ 0, 12, "DEFAULT" ] ],
+        "description": "Lightning arcs towards you!"
+      },
+      [ "MON_LEECH_EVOLUTION", 30 ],
+      [ "PARROT", 40 ]
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]: There is no corresponding speech text for this monster.

File: data/json/monsters/power_leech.json [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    ],
+    "death_drops": {  },
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "NOHEAD", "ELECTRONIC", "IMMOBILE", "NO_BREATHE", "HARDTOSHOOT" ]
```
Comment (Night-Pryanik) [inline data/json/monsters/power_leech.json]: `ELECTRONIC` flag here will result in all sorts of weird behavior, like showing a message of `The root drone sparks and begins searching for a target!` or `The EMP blast fries the root drone!` on using scramblers and emp grenades on it; you'll be able to use robot controls on it; and others. Here and most other monsters.

**Comment by Night-Pryanik** [comment]
Several questions:
1. This new monsters have eyes or some other means of vision? Especially night vision. 30 tiles of night vision is a _very good_ range.
2. Since they have electricity-based attacks, maybe add an electric defense function to them, like shockers have?

File: src/monattack.cpp [hunk 11]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+        !is_adjacent( z, target, false ) ||
+        !z->sees( *target ) ) {
+        return false;
+    }
+    if( !z->can_act() ) {
+        return false;
+    }
+
+    z->moves -= 100;
+    bool uncanny = target->uncanny_dodge();
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
    const bool uncanny = target->uncanny_dodge();
```

File: src/monattack.cpp [hunk 12]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    if( uncanny || dodge_check( z, target ) ) {
+        auto msg_type = target == &g->u ? m_warning : m_info;
+        target->add_msg_player_or_npc( msg_type, _( "The %s lunges at you, but you dodge!" ),
+                                       _( "The %s lunges at <npcname>, but they dodge!" ),
+                                       z->name() );
+        if( !uncanny ) {
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: This will never trigger. This check should be outside of `if( uncanny || dodge_check( z, target ) )` check.

File: src/monattack.cpp [hunk 13]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    if( damage_dealt > 0 ) {
+        auto msg_type = target == &g->u ? m_bad : m_warning;
+        target->add_msg_player_or_npc( msg_type,
+                                       _( "The %1$s impales yor chest for %2$d damage!" ),
+                                       _( "The %1$s impales <npcname>'s chest for %2$d damage!" ),
+                                       z->name(), damage_dealt );
+    } else {
+        target->add_msg_player_or_npc(
+            _( "The %1$s attempts to burrow itself into you, but is stopped by your armor!" ),
+            _( "The %1$s slashes at <npcname>'s %2$s, but is stopped by its glances off armor!" ),
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
            _( "The %1$s slashes at <npcname>'s %2$s, but is stopped by their armor!" ),
```

File: src/monattack.cpp [hunk 14]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+            z->name() );
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: You need to add a parameter here for `%2$s` one line above.

File: src/monattack.cpp [hunk 15]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+        if( !uncanny ) {
+            target->on_dodge( z, z->type->melee_skill * 2 );
+            target->add_msg_player_or_npc( msg_type, _( "The %s lunges at you, but you dodge!" ),
+                                           _( "The %s lunges at <npcname>, but they dodge!" ),
+                                           z->name() );
+        }
+        return true;
+    }
+    const std::string old_name = z->name();
+    const bool could_see_z = g->u.sees( *z );
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: Move these vars into check at line 4777, as they are used only there.

File: src/monattack.cpp [hunk 16]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+        const std::string upgrade_name = z->name();
+        const bool can_see_z_upgrade = g->u.sees( *z );
+        if( could_see_z && can_see_z_upgrade ) {
+            add_msg( m_warning, _( "The %1$s burrows within %2$s corpse and a %3$s emerges from the remains!" ),
+                     old_name,
+                     target_name, upgrade_name );
+        } else if( could_see_z ) {
+            add_msg( m_warning, _( "The %1$s burrows within %2$s corpse!" ), old_name, target_name );
+        } else if( can_see_z_upgrade ) {
+            add_msg( m_warning, _( "A %1$s emerges from %2$s corpse!" ), target->disp_name(), target_name );
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
            add_msg( m_warning, _( "A %1$s emerges from %2$s corpse!" ), upgrade_name, target_name );
```

File: src/monattack.cpp [hunk 17]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+bool mattack::leech_spawner( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    std::list<monster *> allies;
+    for( monster &candidate : g->all_monsters() ) {
+        if( candidate.type->in_species( LEECH_PLANT ) && !candidate.type->has_flag( MF_IMMOBILE ) ) {
+            allies.push_back( &candidate );
+        }
+    }
+    if( allies.size() > 35 ) {
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: 35 or 45, like you said in OP?

File: src/monattack.cpp [hunk 18]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+        return true;
+    }
+    int monsters_spawned = rng( 1, 4 );
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
    const int monsters_spawned = rng( 1, 4 );
```

File: src/monattack.cpp [hunk 19]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+                add_msg( m_warning,
+                         _( "An egg pod ruptures and a %s crawls out from the remains!" ), new_mon->name() );
+            }
+        }
+    }
+    if( one_in( 25 ) ) {
+        z->poly( mon_leech_stalk );
+        if( u_see ) {
+            add_msg( m_warning,
+                     _( "Resplendent fronds emerge from the still intact pods! " ) );
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
                     _( "Resplendent fronds emerge from the still intact pods!" ) );
```

File: src/monattack.cpp [hunk 20]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+}
+
+bool mattack::mon_leech_evolution( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    const bool is_queen = z->type->has_flag( MF_QUEEN );
+    std::list<monster *> queens;
+    for( monster &candidate : g->all_monsters() ) {
+        if( candidate.type->in_species( LEECH_PLANT ) && candidate.type->has_flag( MF_QUEEN ) &&
+            rl_dist( z->pos(), candidate.pos() ) ) {
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: What do you want to accomplish here with `rl_dist`?

File: src/monattack.cpp [hunk 21]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    if( !is_queen ) {
+        if( queens.empty() ) {
+            z->poly( mon_leech_blossom );
+            z->set_hp( z->get_hp_max() );
+            if( u_see ) {
+                add_msg( m_warning,
+                         _( "The %s blooms into flowers!" ), z->name() );
+            }
+        }
+
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
```

File: src/monattack.cpp [hunk 22]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    } else {
+        if( !queens.empty() ) {
+            if( u_see ) {
+                add_msg( m_warning,
+                         _( "The %s flowers whiter and fall!" ), z->name() );
+            }
+            z->poly( mon_leech_stalk );
+            g->m.spawn_item( z->pos(), "leech_flower", 5, 0, calendar::turn );
+
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
```

File: src/monattack.cpp [hunk 23]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+        }
+    }
+    return true;
+}
+
+
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
```

File: src/monattack.cpp (hunk 12 above)
Comment (BevapDin) [reply src/monattack.cpp]: Why not? It will be triggered when `uncanny` is false, but ` dodge_check( z, target )` yields true.

However: the message about lunging and dodging will be shown twice (once above and again inside this branch).

File: src/monattack.cpp [hunk 24]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    }
+    return true;
+}
+
+bool mattack::leech_spawner( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    std::list<monster *> allies;
+    for( monster &candidate : g->all_monsters() ) {
+        if( candidate.type->in_species( LEECH_PLANT ) && !candidate.type->has_flag( MF_IMMOBILE ) ) {
```
Comment (BevapDin) [inline src/monattack.cpp]:
```suggestion
        if( candidate.in_species( LEECH_PLANT ) && !candidate.has_flag( MF_IMMOBILE ) ) {
```
Those functions exist as part of the `monster` class, so you don't need to check for them within the monster type.

File: src/monattack.cpp [hunk 25]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+    return true;
+}
+
+bool mattack::mon_leech_evolution( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    const bool is_queen = z->type->has_flag( MF_QUEEN );
+    std::list<monster *> queens;
+    for( monster &candidate : g->all_monsters() ) {
+        if( candidate.type->in_species( LEECH_PLANT ) && candidate.type->has_flag( MF_QUEEN ) &&
```
Comment (BevapDin) [inline src/monattack.cpp]:
```suggestion
        if( candidate.in_species( LEECH_PLANT ) && candidate.has_flag( MF_QUEEN ) &&
```

File: src/monattack.cpp [hunk 26]
```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
...
+                     _( "Resplendent fronds emerge from the still intact pods! " ) );
+        }
+    }
+    return true;
+}
+
+bool mattack::mon_leech_evolution( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    const bool is_queen = z->type->has_flag( MF_QUEEN );
```
Comment (BevapDin) [inline src/monattack.cpp]:
```suggestion
    const bool is_queen = z->has_flag( MF_QUEEN );
```

File: src/monster.cpp [hunk 27]
```diff
@@ -317,6 +317,12 @@ int monster::get_upgrade_time() const
     return upgrade_time;
 }
 
+// Sets time to upgrade to 0.
+void monster:: allow_upgrade()
```
Comment (BevapDin) [inline src/monster.cpp]:
```suggestion
void monster::allow_upgrade()
```

File: data/json/items/comestibles/carnivore.json (hunk 1 above)
Comment (John-Candlebury) [reply data/json/items/comestibles/carnivore.json]: The other tainted pieces of meat warn you that eating them is a poor idea, but you are right, I'll reword it to sound less meta-gamey.

File: data/json/monsters/power_leech.json (hunk 10 above)
Comment (John-Candlebury) [reply data/json/monsters/power_leech.json]: Robot control works based on robot species,  and they being weak to emp is intended, but you are right about the scrambler grenade, I'll change it.

**Comment by alexanderglasco-ui** [comment]
Pretty awful as a game design choice-  essentially, if you see these, you will die, unless you happen to have a flamethrower on your person.

There's pretty much the only mob in the game you can't reasonably completely run from and can't handle with normal firearms or kiting, that can be in random rural areas.

John's pedantic' suggest a flamethrower or explosives' implies the player MUST have, equip, and fuel, a flamethrower ON THEIR PERSON at ALL TIMES.

I know lots of people save scum, but I don't know why you'd make content that forces you to save scum. There's definitely multiple camps of devs, and John falls into the 'haha won't this be an interesting and satisfying way to die, so you can scum-puzzle it out?' camp.

**Comment by RenechCDDA** [comment]
@alexanderglasco-ui Review the code of conduct before posting again, you can consider this an official warning.
